aiosmtpd/handlers.py

```diff
@@ -155,6 +155,29 @@ def handle_message(self, message):
 
 
 @public
+class AsyncMessage(Message):
+
+    @asyncio.coroutine
+    def process_message(self, peer, mailfrom, rcpttos, data, *, loop, **kws):
+        # If the server was created with decode_data True, then data will be a
+        # str, otherwise it will be bytes.
+        if isinstance(data, bytes):
+            message = message_from_bytes(data, self.message_class)
+        else:
+            assert isinstance(data, str), (
+              'Expected str or bytes, got {}'.format(type(data)))
+            message = message_from_string(data, self.message_class)
+        message['X-Peer'] = str(peer)
+        message['X-MailFrom'] = mailfrom
+        message['X-RcptTos'] = COMMASPACE.join(rcpttos)
+        yield from self.handle_message(message, loop=loop)
```

Comment:
Nearly all of this is shared with the `Message.process_message()` base class method.  Is there a good way to refactor the code so that it doesn't have to be duplicated?  That would make it much easier if we ever have to change something (e.g. add another default `X-` header.